### ISSUE TITLE
README fix needed

### ISSUE THREAD

**Issue by CarlKCarlK** [matgla/Renode_RP2040#1]
I think

`(monitor) include @repos/Renode_RP2040/board/initialize_raspberry_pico.resc`

should be

`(monitor) include @repos/Renode_RP2040/boards/initialize_raspberry_pico.resc`

--------

I love the idea of this project. I'm trying to get it running on Windows. I was having problems related to mono/dotnet. I may have fixed those. Now I'm getting:

```
P:\pico\blinky_carl2>renode --console
19:06:38.7553 [INFO] Loaded monitor commands from: C:\Program Files\Renode\scripts/monitor.py
Renode, version 1.15.3.8708 (fb0ff392-202411250449)

(monitor) path add @c:/deldir/1124/Renode_RP2040
Current 'PATH' value is: C:\Program Files\Renode;P:\pico\blinky_carl2;c:\deldir\1124\Renode_RP2040
(monitor) include @c:/deldir/1124/Renode_RP2040/boards/initialize_raspberry_pico.resc
19:06:52.7434 [INFO] Including script: c:\deldir\1124\Renode_RP2040\boards\initialize_raspberry_pico.resc
19:06:52.7769 [INFO] System bus created.
19:06:55.9235 [INFO] Including script: c:\deldir\1124\Renode_RP2040\initialize_common.resc
Errors during compilation or loading:
Could not compile assembly: c:\deldir\1124\Renode_RP2040\emulation\peripherals\gpio\rp2040_gpio.cs(163,30): Antmicro.Renode.Peripherals.GPIOPort.RP2040GPIO.GetFunction(int)
(raspberry_pico) 
```
My goal is to run a simple project with one button and two LEDs.

I'll watch the project and look forward to its progress. - Carl)

**Comment by matgla** [comment]
Hi, 
Thank you for reporting that issue!

I never used the Windows version of Renode, and I found an issue where C# array initialization doesn't work while using the Windows .dotnet version. 

I am preparing a workaround to unblock your testing :) 
After that, I'll report the issue to Renode.

I'll keep you informed when I successfully run Renode with my addons on the Windows platform. 

About the readme typo, of course, it should be boards. I'll fix that. 

Thanks :)

**Comment by matgla** [comment]
And there is one more problem. 
Renode doesn't support the binding of native Windows DLLs for external libraries. 
So, my PIO emulator must be compiled the same as the Verilator modules described: 
https://renode.readthedocs.io/en/latest/tutorials/co-simulating-custom-hdl.html 

My emulator continually compiles the Pio emulator library on its own since I cannot currently provide redistributable libraries, which enforces the use of msys with mingw32-make and the GCC toolchain. 

For some reason, even a simple shared library crashes on my msys installation. 

Until I'll solve that maybe you can use Renode under WSL?

**Comment by CarlKCarlK** [comment]
Thanks! I took the Windowfix branch a few minutes ago and things are better.

(I'm not sure if I'll keep perusing emulation, but I thought I'd let you know what it looks like on my side. My goal is to have people be able to run an emulated two LED, one button setup. I'm working on an [article/project](https://medium.com/@carlmkadie) with Rust and embedded and thought it would be cool if people without hardware could run it.)

```
P:\pico\blinky_carl2>renode --console
07:31:57.7803 [INFO] Loaded monitor commands from: C:\Program Files\Renode\scripts/monitor.py
Renode, version 1.15.3.8708 (fb0ff392-202411250449)

(monitor) path add @c:/deldir/1124/Renode_RP2040
Current 'PATH' value is: C:\Program Files\Renode;P:\pico\blinky_carl2;c:\deldir\1124\Renode_RP2040
(monitor) include @c:/deldir/1124/Renode_RP2040/boards/initialize_raspberry_pico.resc
07:32:30.9377 [INFO] Including script: c:\deldir\1124\Renode_RP2040\boards\initialize_raspberry_pico.resc
07:32:30.9592 [INFO] System bus created.
07:32:40.7576 [INFO] Including script: c:\deldir\1124\Renode_RP2040\initialize_common.resc (3)
07:32:41.8888 [WARNING] Using alias 'id' for parameter 'cpuId' (4)
07:32:42.1989 [INFO] [no-name]: Setting ROSC frequency to: 6MHz
07:32:42.2189 [ERROR] Binding
07:32:42.2239 [ERROR] BINDDED
07:32:42.2279 [ERROR] Binding
07:32:42.2289 [ERROR] BINDDED
07:32:42.3014 [WARNING] Transmission finished in unexpected state: RecognizeOperation
07:32:42.4895 [INFO] sysbus: Loaded SVD: C:\Users\carlk\AppData\Local\Temp\renode-2106764\558c156c-1fae-4a56-a70d-31c6a5689e17.tmp. Name: RP2040. Description: 
        Dual-core Arm Cortex-M0+ processor, flexible clock running up to 133 MHz\n
        264KB on-chip SRAM\n
        2 x UART, 2 x SPI controllers, 2 x I2C controllers, 16 x PWM channels\n
        1 x USB 1.1 controller and PHY, with host and device support\n
        8 x Programmable I/O (PIO) state machines for custom peripheral support\n
        Supported input power 1.8-5.5V DC\n
        Operating temperature -20C to +85C\n
        Drag-and-drop programming using mass storage over USB\n
        Low-power sleep and dormant modes\n
        Accurate on-chip clock\n
        Temperature sensor\n
        Accelerated integer and floating-point libraries on-chip
      .
07:32:42.7525 [INFO] sysbus: Loading block of 16384 bytes length at 0x0.
07:32:42.7660 [INFO] sysbus: Loading block of 3052 bytes length at 0x50100400.
07:32:42.7850 [INFO] piocpu0: Setting PC value to 0x0.
07:32:42.7850 [INFO] piocpu1: Setting PC value to 0x0.
(raspberry_pico) include @tests/pi_pico_with_leds_and_button.resc
07:32:48.0155 [INFO] Including script: P:\pico\blinky_carl2\tests\pi_pico_with_leds_and_button.resc
07:32:48.0175 [INFO] System bus created.
There was an error executing command 'emulation AddExternalConnector "LED0" GPIO'
emulation does not provide a field, method or property AddExternalConnector.
```
I assume the last errors are because I don't know what I'm doing and I'm just using ChatGPT4o to try to write my *.resc code:

```
using sysbus
mach create "pi_pico"

# Load RP2040 and Raspberry Pi Pico platform definitions
# machine LoadPlatformDescription @c:/deldir/1124/Renode_RP2040/rp2040.repl
# machine LoadPlatformDescription @c:/deldir/1124/Renode_RP2040/raspberry_pico.repl

# Define GPIOs for LEDs
emulation AddExternalConnector "LED0" GPIO
emulation AddExternalConnector "LED1" GPIO
connector Connect sysbus.gpio2 LED0
connector Connect sysbus.gpio3 LED1

# Define GPIO for Button
emulation AddExternalConnector "Button" GPIO
connector Connect sysbus.gpio13 Button

# Attach visualizations
emulation AddGPIOVisualizer LED0
emulation AddGPIOVisualizer LED1
emulation AddGPIOVisualizer Button

# Load Firmware
macro reset
"""
    sysbus LoadBinary @target/thumbv6m-none-eabi/debug/blinky 0x10000000
    start
"""
runMacro reset
```

I think RENODE and your rp2040 projects are great.

**Comment by matgla** [comment]
Thanks, for so fast testing :) 
I am still fighting with GitHub CI to enable regression for Windows too :)

I think it may be better to define your platform description, it's pretty intuitive. 

You can look inside: https://github.com/matgla/Renode_RP2040/blob/main/tests/pio/pio_blink/raspberry_pico_with_more_leds.repl

How to use it is inside: 
https://github.com/matgla/Renode_RP2040/blob/main/tests/pio/pio_blink/pio_blink.resc

This test contains board definition with 3 additional LEDs. 
Then I see inside renode repository that they have also a button simulator implemented -> Miscellaneous.Button. 

But I am not sure if I already implemented connector inputs. I haven't tested that button peripheral. 

To be honest, GPIO is not fully implemented yet, thus yellow color. For example, IRQ propagation to NVIC is not yet possible. 

I think I can add a test with buttons and LEDs, but not today. 
Maybe still during this week :) 

One note: 
Miscellaneous.LED is printing its states on noisy log level, so you should keep `logLevel -1` in your *.resc file or enable it each execution before starting.

**Comment by matgla** [comment]
Maybe it will work: 
I've added one simple fix to Windowfix branch and I just only tested if simulation sees button pressing. 

```
renode  .\tests\button\button.resc

(pico_tests) logLevel -1
s
Starting emulation...
(pico_tests) sysbus.gpio.button Press
(pico_tests) sysbus.gpio.button Release
(pico_tests) sysbus.gpio.button Press
```

Shows that GPIO state was indeed:
```
18:03:47.5279 [NOISY] gpio: Setting GPIO4 to: False, time: 00:00:25.457130000, from: NONE
18:03:47.9953 [NOISY] gpio: Setting GPIO4 to: True, time: 00:00:25.571040000, from: NONE
```
So if there is no bugs in state management firmware should be able to read GPIO state (except you are using interrupts to detect edges on GPIO, then my simulation is not mature enough).

I think you can adapt platform description and try using with your firmware. 

In the meantime, I will prepare my firmware to write a real test using buttons and LEDs.

**Comment by CarlKCarlK** [comment]
If you want to see the project that I'd love to emulate it is here:
https://github.com/CarlKCarlK/blinky_carl2
(Elf file: https://github.com/CarlKCarlK/blinky_carl2/releases/tag/0.0.1-alpha)

It does look for a falling edge on the button signal.

**Comment by matgla** [comment]
Sure, it's going to be interesting to run your example. 
Having binary it will be to be much easier :) 

I checked, and the Rust environment uses an additional register, which still needs to be added, but that's trivial. 
Adding NVIC interrupts for GPIO is more complex, and embassy-rs are using them, so I will prioritize that topic. 
I'll continue tomorrow and let you know when I execute your binary correctly.

**Comment by matgla** [comment]
Hi, 
I still have to stabilize a few things, but you should be able to run your firmware successfully using the latest master. 

I tested it already using the below method:

`raspberry_blinky.repl`

```
using "rp2040.repl"

led1: Miscellaneous.LED @ gpio 2
led2: Miscellaneous.LED @ gpio 3

gpio: 
    2 -> led1@0
    3 -> led2@0

button: Miscellaneous.Button @ gpio 13 
    -> gpio@13

```

`run_blinky.resc`
```
$platform_file=@raspberry_blinky.repl
$global.FIRMWARE=@blinky
include @run_firmware.resc

logLevel -1 sysbus.gpio.led1
logLevel -1 sysbus.gpio.led2

```

then to use button inside renode monitor:
```
(rasbperry_pico) sysbus.gpio.button Press
(rasbperry_pico) sysbus.gpio.button Release
(rasbperry_pico) sysbus.gpio.button Press
(rasbperry_pico) sysbus.gpio.button Press
(rasbperry_pico) sysbus.gpio.button Release
(rasbperry_pico) 
```

In the log, LED state changes should be visible, precisely like on real hardware. 

Let me know if it works for you! After confirmation, I can release a tag for you to keep this version in history since further ones may break some scripting interfaces.

**Comment by CarlKCarlK** [comment]
Excellent! I've been having fun watching all the checkins. I'll be able to check it within a couple of hours.

**Comment by matgla** [comment]
One little tip: you may want to see simulated time changes. 
Unfortunately, the LED doesn't print that, but you can also enable logs on GPIO with `logLevel -1 sysbus.gpio`, and the time printed on the GPIO state change is simulation time.

```
21:05:28.5672 [NOISY] gpio: Setting GPIO3 to: True, time: 00:00:52.038200000, from: SIO
21:05:28.5672 [NOISY] gpio.led2: LED state changed to True
21:05:28.5681 [NOISY] gpio: Setting GPIO2 to: True, time: 00:00:52.038500000, from: SIO
21:05:28.5681 [NOISY] gpio.led1: LED state changed to True
21:05:28.9378 [NOISY] gpio: Setting GPIO3 to: False, time: 00:00:52.238400000, from: SIO
21:05:28.9379 [NOISY] gpio.led2: LED state changed to False
21:05:28.9388 [NOISY] gpio: Setting GPIO2 to: False, time: 00:00:52.238700000, from: S
```
I discovered that the PressAndRelease function from Button is too fast for bouncy suspension. The time between pressing and releasing from the renode monitor needs to be quick enough; otherwise, the code goes to 'SOS state'. 

It may be wise to write a script in *.resc that iterates toward states using the button.

**Comment by CarlKCarlK** [comment]
I think I need some help knowing how to tie things together. Here is what I'm doing now:

* sync the windowsFix branch. That lives in C:\deldir\1124\Renode_RP2040
* In my blinky_carlk2 project:
```
cd tests
renode --console
path add @c:/deldir/1124/Renode_RP2040
include @c:/deldir/1124/Renode_RP2040/boards/initialize_raspberry_pico.resc
include @run_blinky.resc
```
where `tests/run_blinky.resc` is

```
$platform_file=@raspberry_blinky.repl
$global.FIRMWARE=@../target/thumbv6m-none-eabi/debug/blinky
include @run_firmware.resc

logLevel -1 sysbus.gpio.led1
logLevel -1 sysbus.gpio.led2
```

and `tests/raspberry_blinky.repl` is
```
using "rp2040.repl"

led1: Miscellaneous.LED @ gpio 2
led2: Miscellaneous.LED @ gpio 3

gpio: 
    2 -> led1@0
    3 -> led2@0

button: Miscellaneous.Button @ gpio 13 
    -> gpio@13
```

As you can tell, I really don't know what's going on with RENODE, *.repl files, *.resc files etc.
Any guidance would be much appreciated.

-- Carl

**Comment by CarlKCarlK** [comment]
It's working! Thanks! Below is my `run_blinky.resc`.

Is this running in real time? Do you know how I can call these two commands:
```
sysbus.gpio.button Press
sysbus.gpio.button Release
```
with a set time between them (or no time)? The firmware distinguishes between long and short taps on the button and currently everything I do is a long tap.

-- Carl

```
$platform_file=@raspberry_blinky.repl
$global.FIRMWARE=@../target/thumbv6m-none-eabi/debug/blinky
path add @c:/deldir/1124/Renode_RP2040
include @run_firmware.resc

logLevel -1 sysbus.gpio
logLevel -1 sysbus.gpio.led1
logLevel -1 sysbus.gpio.led2
```

**Comment by matgla** [comment]
I don't know how to do it properly, but one idea came to mind. 
Just add another script, `short_tap.resc`: 
```
sysbus.gpio.button Press
sleep 0.2
sysbus.gpio.button Release
```

And then include it from the Renode console: 
```
i @short_tap.resc
```

**Comment by CarlKCarlK** [comment]
This worked: `sysbus.gpio.button Press; sysbus.gpio.button Release`.
Has anyone connected a little GUI to RENODE? It would be great to have a button to click and leds to see.

**Comment by matgla** [comment]
Unfortunately, that's a question for the Antmicro/Renode team. I am not related to Antmicro company. 
When I was looking for a GUI, I found nothing useful. There is some visualization described there, but it is most probably not what you are looking for: 
https://github.com/antmicro/renode-board-visualization 

I think that Renode is focused on automated tests. 

It is certainly possible to spawn GUI from other processes and communicate through some socket. But that's a task for the future, especially to add some displays. 
For now, from my side, more critical things need to be implemented to improve the coverage of simulated hardware.

**Comment by matgla** [comment]
Hi, 
I released stabilized tag v0.1.0. 
Please update your scripts if needed to follow the new path layout.
Boards are now in `/boards` directory and cores will be in `/cores`. 

I am closing that issue since `blinky` is working. If a new problem appears, please create a new issue!

Best regards,
Mateusz

**Comment by matgla** [comment]
@CarlKCarlK: FYI, I just released v0.2.0 with proof of concept of Raspberry Pico board visualization. 
If you wish to try it please follow the readme. 
I know it's ugly, but it seems it is working quite well.

**Comment by CarlKCarlK** [comment]
Cool! I'll take a look in the next couple of hours.

**Comment by CarlKCarlK** [comment]
I'm getting a 
```
(rasbperry_pico) startVisualization 8080
Spawned process with PID: 925932
(rasbperry_pico) Process IO has died
```
Might be a Windows problem.

Here are the steps I took:

* Sync to your main branch (last update 2 hours ago)
* `pip3 install -r C:\deldir\1124\Renode_RP2040\visualization\requirements.txt`

says:
```
...
Requirement already satisfied: idna>=2.0 in c:\users\carlk\appdata\roaming\python\python312\site-packages (from yarl<2.0,>=1.17.0->aiohttp==3.11.8->-r C:\deldir\1124\Renode_RP2040\visualization\requirements.txt (line 3)) (3.10)
```
Add `cores/` to my `raspberry_blinky.repl`:

```
using "cores/rp2040.repl"

led1: Miscellaneous.LED @ gpio 2
led2: Miscellaneous.LED @ gpio 3

gpio: 
    2 -> led1@0
    3 -> led2@0

button: Miscellaneous.Button @ gpio 13 
    -> gpio@13
```

My `run_blinky.resc` remains:
```
$platform_file=@raspberry_blinky.repl
$global.FIRMWARE=@../target/thumbv6m-none-eabi/debug/blinky
path add @c:/deldir/1124/Renode_RP2040
include @run_firmware.resc

logLevel -1 sysbus.gpio
logLevel -1 sysbus.gpio.led1
logLevel -1 sysbus.gpio.led2
```

* run `renode --console run_blinky.rescs
* I can now do "s" and "p" to start and pause the simulation.
* (rasbperry_pico) startVisualization 8080
Spawned process with PID: 937172
(rasbperry_pico) Process IO has died

**Comment by matgla** [comment]
Hi, 
My Windows machine is working fine. But indeed I had some problems with pip and python3 interworking. 
I added better logging to `main` branch. 

You can also try running the visualization server with: `python3 .\visualization\visualization_server.py --port 1234`. 
If it reports a lack of imported libraries, then for my Windows machine: `python3 -m pip install -r visualization/requirements.txt` was helpful.

**Comment by CarlKCarlK** [comment]
Mateusz,

It's working! (I think the problem related to me have many versions of Python.)
My co-author will be very excited that people without hardware can run this. We may also try to figure out using RENODE in CI because testing of Rust on embedded systems is currently very weak.

How hard is it to set up a new visualization? Here is our other project for the article: a 4-digit 7-segment LED clock:
https://github.com/CarlKCarlK/clock/ (I've added a *.elf release)

I'm also comfortable with Python (and can get by with JavaScript) if that is all that is needed.

-- Carl

**Comment by matgla** [comment]
Adding 7-segment displays shouldn't be complex. However, I am still figuring out how to do it right to support both the robot framework for regression tests and UI. 

First, I have to do something with my JS and Python server to make it more extendable and user-friendly. 
The current UI is just an experiment if the whole idea has any chance of success. 

I created a new topic to track 7-segment display implementation: https://github.com/matgla/Renode_RP2040/issues/7
